README.md
# @abbeymart/res-messages

res-message package includes the standard responses for various activities/operations

## Installation

```sh
$ npm install @abbeymart/res-messages
```

## API

```js
// Coming soon

const abc = 'test';
console.log('abc: ', abc);

```

## Example

```js
// Coming soon

const abc = 'test';
console.log('abc: ', abc);


```

### [MIT Licensed]


HISTORY.md
# Historical logs

## Version 0.2.1
- Released April 06, 2020
- resolved npm security alert issues (minimist, lodash and mem)

## Version 0.2.0
- Released June 15, 2019
- re-factored res-msg function and restructure res-msg meta data

## Version 0.1.1
- Released February 03, 2019
- aligned with mconnect-library version

## Version 0.1.0
- Released January 10, 2019
- Includes standards functions for consistently managing response messages



test/assert.js
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2019-01-03 | @Updated: 2019-01-03
 * @Company: mConnect.biz | @License: MIT
 * @Description: @mconnect/test | assertion function(s)
 */

// test handler function | preferred - elegant/flexible - for handling different cases
function ok( expr, msg ) {
    if ( ! expr ) {
        throw new Error(msg);
    }
}
module.exports = ok;


test/tests.success.js
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2019-01-03 | @Updated: 2019-01-03
 * @Company: mConnect.biz | @License: MIT
 * @Description: @mconnect/res-messages testing, success
 */

const suite             = require('mocha').suite;
const test              = require('mocha').test;
const before            = require('mocha').before;
const HTTPStatusCode    = require('http-status-codes');
const { getResMessage } = require('../index');
const ok                = require('./assert');

let res,
    msgType = '',
    options = {};

before(() => {
    msgType = 'success';
    options = {
        value  : '',
        code   : '',
        resCode: '',
        message: '',
    };
    res     = {
        code   : 'success',
        resCode: HTTPStatusCode.OK,
        value  : '',
        message: 'Request completed successfully',
    };
});

suite('@mconnect/res-messages package Testing - success:', () => {
    suite('Positive testing:', () => {
        test('should return success code for success-message', () => {
            const req = getResMessage(msgType, options);
            ok(res.code === req.code, 'response-code should be: success');
        });
        test('should return ok/200 resCode for success-message', () => {
            const req = getResMessage(msgType);
            ok(res.resCode === req.resCode, 'resCode should be: OK/200');
        });
        test('should return Completed successfully message for success-message', () => {
            const req = getResMessage(msgType);
            ok(res.message === req.message, 'message should be: Completed successfully');
        });
    });

    suite('Negative testing:', () => {
        test('should return ok/200 resCode for success-message', () => {
            res       = {
                code   : 'unauthorised',
                resCode: HTTPStatusCode.OK,
                value  : '',
                message: 'Request completed successfully.',
            };
            const req = getResMessage(msgType);
            ok(res.code !== req.code, 'response-code should be: success, not unauthorised');
        });
    });
});

test/tests.checkError.js
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2019-01-03 | @Updated: 2019-01-03
 * @Company: mConnect.biz | @License: MIT
 * @Description: @mconnect/res-messages testing: checkError
 */

const suite             = require('mocha').suite;
const test              = require('mocha').test;
const before            = require('mocha').before;
const HTTPStatusCode    = require('http-status-codes');
const { getResMessage } = require('../index');
const ok                = require('./assert');

let res,
    msgType = '',
    options = {};


before(() => {
    msgType = 'checkError';
    options = {
        value  : '',
        code   : '',
        resCode: '',
        message: '',
    };
    res     = {
        code   : 'paramsError',
        resCode: HTTPStatusCode.NOT_ACCEPTABLE,
        value  : '',
        message: 'Parameters checking error',
    };
});

suite('@mconnect/res-messages package Testing - checkError:', () => {
    suite('Positive testing:', () => {
        test('should return paramsError code for checkError-message', () => {
            const req = getResMessage(msgType, options);
            ok(res.code === req.code, 'response-code should be: paramError');
        });
        test('should return NOT_ACCEPTABLE/406 resCode', () => {
            const req = getResMessage(msgType);
            ok(res.resCode === req.resCode, 'resCode should be: NOT_ACCEPTABLE/406');
        });
        test('should return Parameters checking error message', () => {
            const req = getResMessage(msgType);
            ok(res.message === req.message, 'message should be: Parameters checking error');
        });
    });

    suite('Negative testing:', () => {
        test('should return paramError for checkError-message', () => {
            res       = {
                code   : 'unauthorised',
                resCode: HTTPStatusCode.OK,
                value  : '',
                message: 'Parameters checking error',
            };
            const req = getResMessage(msgType);
            ok(res.code !== req.code, 'response-code should be: paramError, not unauthorised');
        });
    });
});


src/resMessages.js
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2019-01-03 | @Updated: 2019-06-15
 * @Company: mConnect.biz | @License: MIT
 * @Description: @mconnect/res-messages, response-messages | utility functions
 */

const resMsgParams = require('./resMsgParams');

module.exports = {
    getResMessage(msgType = '', options = {}) {
        let value   = '',
            code    = '',
            resCode = '',
            message = '';

        if (msgType && resMsgParams[msgType]) {
            code    = resMsgParams[msgType].code;
            value   = options && options.value ? options.value : resMsgParams[msgType].value;
            resCode = options && options.resCode ? options.resCode : resMsgParams[msgType].resCode;
            message = options && options.message ? options.message : resMsgParams[msgType].message;
        } else if (msgType && resMsgParams[msgType] === undefined) {
            code    = msgType;
            value   = options && options.value ? options.value : resMsgParams['stdRes'].value;
            resCode = options && options.resCode ? options.resCode : resMsgParams['stdRes'].resCode;
            message = options && options.message ? options.message : resMsgParams['stdRes'].message;
        } else {
            // resMsgParams for default response
            value   = options && options.value ? options.value : resMsgParams['stdRes'].value;
            code    = options && options.code ? options.code : resMsgParams['stdRes'].code;
            resCode = options && options.resCode ? options.resCode : resMsgParams['stdRes'].resCode;
            message = options && options.message ? options.message : resMsgParams['stdRes'].message;
        }
        return msgFunc(value, code, resCode, message);
    },
};

function msgFunc(value, code, resCode, msg) {
    return {
        code   : code,
        resCode: resCode,
        value  : value,
        message: msg,
    };
}


src/resMsgParams.js
/**
 * @Author: abbeymart | Abi Akindele | @Created: 2019-01-06 | @Updated: 2019-05-15
 * @Company: mConnect.biz | @License: MIT
 * @Description: @mconnect/res-messages, response-messages | settings, default values
 */

const HTTPStatusCode = require('http-status-codes');

// message options => code, resCode, message, value
const resMsgParams = {
    checkError: {
        code   : 'paramsError',
        resCode: HTTPStatusCode.NOT_ACCEPTABLE,
        message: 'Parameters checking error',
        value  : '',
    },
    connectError: {
        code   : 'connectionError',
        resCode: HTTPStatusCode.NETWORK_AUTHENTICATION_REQUIRED,
        message: 'Connection error',
        value  : '',
    },
    validateError: {
        code   : 'paramsError',
        resCode: HTTPStatusCode.NOT_ACCEPTABLE,
        message: 'Validation error for inputs parameters',
        value  : '',
    },
    tokenExpired: {
        code   : 'tokenExpired',
        resCode: HTTPStatusCode.UNAUTHORIZED,
        message: 'Unauthorized. Token / Access-key has expired. Please login again',
        value  : '',
    },
    unAuthorized: {
        code   : 'unAuthorized',
        resCode: HTTPStatusCode.UNAUTHORIZED,
        message: 'Unauthorised Action or Task',
        value  : '',
    },
    notFound: {
        code   : 'notFound',
        resCode: HTTPStatusCode.NOT_FOUND,
        message: 'Requested information not found',
        value  : '',
    },
    success: {
        code   : 'success',
        resCode: HTTPStatusCode.OK,
        message: 'Request completed successfully',
        value  : '',
    },
    removeDenied: {
        code   : 'removeDenied',
        resCode: HTTPStatusCode.UNAUTHORIZED,
        message: 'Remove action/task denied/unauthorised',
        value  : '',
    },
    removeError: {
        code   : 'removeError',
        resCode: HTTPStatusCode.NOT_MODIFIED,
        message: 'Error removing/deleting information, retry or contact system-admin',
        value  : '',
    },
    removed: {
        code   : 'removed',
        resCode: HTTPStatusCode.OK,
        message: 'Record(s) deleted/removed successfully',
        value  : '',
    },
    subItems: {
        code   : 'subItems',
        resCode: HTTPStatusCode.NOT_MODIFIED,
        message: 'Record includes sub-items, which must be removed first',
        value  : '',
    },
    duplicateRec: {
        code   : 'duplicate',
        resCode: HTTPStatusCode.NOT_MODIFIED,
        message: 'Duplicate record exists',
        value  : '',
    },
    updated: {
        code   : 'updated',
        resCode: HTTPStatusCode.OK,
        message: 'Information update action completed successfully',
        value  : '',
    },
    updateError: {
        code   : 'updateError',
        resCode: HTTPStatusCode.NOT_MODIFIED,
        message: 'Error updating information/record(s)',
        value  : '',
    },
    updateDenied: {
        code   : 'updateDenied',
        resCode: HTTPStatusCode.UNAUTHORIZED,
        message: 'Update action/task not authorised',
        value  : '',
    },
    inserted: {
        code   : 'inserted',
        resCode: HTTPStatusCode.OK,
        message: 'Information/record(s) inserted/created successfully',
        value  : '',
    },
    insertError: {
        code   : 'insertError',
        resCode: HTTPStatusCode.NOT_MODIFIED,
        message: 'Error inserting/creating new information/record',
        value  : '',
    },
    recExist: {
        code   : 'exists',
        resCode: HTTPStatusCode.NOT_MODIFIED,
        message: 'Document/record exists',
        value  : '',
    },
    stdRes    : {
        code   : 'unknown',
        resCode: HTTPStatusCode.UNPROCESSABLE_ENTITY,
        message: 'Unspecified response/error message',
        value  : '',
    },
};

module.exports = resMsgParams;

